### ISSUE TITLE
Wavering render since Stream deprecation

### ISSUE THREAD

**Issue by ThomasGorisse** [google/filament#5577]
Hi,

We have some rendering issues since moving from `Stream` to `Texture.import()`.
1.20.0 was rendering quite great with the new system even if we still got some green screen on some frames.
But the 1.22.1 seems to introduce some wave effects on our Camera stream coming from ARCore.
| 1.20.0 | 1.22.1 |
| - | - |
| https://user-images.githubusercontent.com/6597529/169303162-66bf6b29-4d4f-44cc-85af-38b2895062e4.mp4 | https://user-images.githubusercontent.com/6597529/169302996-c6fc05c8-104a-4586-bf2a-9b27a8fe18d2.mp4 |

- *Tested on Pixel 4a*
- *The positioning issue might come from ARCore latest version*

Here is how we use it:
https://github.com/SceneView/sceneview-android/blob/3ad435fdc0f2578e9b37bb0778182d45d1a50a31/arsceneview/src/main/java/io/github/sceneview/ar/camera/ArCameraStream.kt#L225-L256

https://github.com/SceneView/sceneview-android/blob/842e2d08307b50ebca5832a0974dd0865674cd5a/sceneview/src/main/java/com/google/ar/sceneform/rendering/GLHelper.java#L54-L66

Thanks

**Comment by pixelflinger** [comment]
It's pretty obvious from the artifacts that you're not selecting the right texture. It looks like the texture that is being sampled from is also one that is being written to by camera.
We (some teams at google) are using importTexture internally with ARCore with success.

**Comment by ThomasGorisse** [comment]
Thanks for your answer.

@grassydragon Does that make sense for you?

@pixelflinger Are you thinking of one day making public the internal work you made around ARCore and Filament because the Android community would really need it.

And as much as we can work on a framework trying to make AR accessible, since it's not a full time job for any of us, it's quite hard to follow every ARCore and Filament updates and bring new features at same time (and answering on Discord)

It would be great to see that AR is officially possible for Android developers without Unity or knowing OpenGL.

**Comment by romainguy** [comment]
>  Are you thinking of one day making public the internal work you made around ARCore and Filament because the Android community would really need it.

Everything that Filament does is available publicly.

**Comment by grassydragon** [comment]
> Does that make sense for you?

Yes, it can be, however, I thought we would see a place where the texture is torn in too parts. Anyway, I need to get familiar with the new way of working with textures.

**Comment by pixelflinger** [comment]
I'm going to close this for now -- reopen if you don't find the cuprit.

**Comment by grassydragon** [comment]
We are now trying to solve this issue:

> The positioning issue might come from ARCore latest version

It turns out that this is an issue with synchronization of the 3D scene state and camera frames.

Synchronization in the Filament `Stream` with an external texture was handled automatically by making a texture copy that corresponds to the frame being rendered: https://github.com/google/filament/blob/dab02e915e9ec0bc8dfe4436debcdd6e420c9f75/android/filament-android/src/main/java/com/google/android/filament/Stream.java#L69

However, when using multiple textures with ARCore how can we know when we can advance the texture index? The `FrameCompletedCallback` works only for Metal: https://github.com/google/filament/blob/347642985f7a796303c97602010b981975a40516/android/filament-android/src/main/java/com/google/android/filament/SwapChain.java#L121

Currently, we are selecting the next texture before scheduling each frame for rendering (calling the `Renderer.render()` method): https://github.com/SceneView/sceneview-android/blob/f00bcf5d04edc2977c42e36968c02aabda31b4bd/arsceneview/src/main/java/io/github/sceneview/ar/camera/ArCameraStream.kt#L254

**Comment by ThomasGorisse** [comment]
It's fixed.
https://user-images.githubusercontent.com/6597529/174429857-e4fa8559-a033-4bc9-a88c-4a0055a9925a.mp4
We will now try to find the best engine config to improve rendering (We use the default engine configuration for now)

Thanks for your answer.

**Comment by ss9326** [comment]
Hi, I'm facing the same problem using ARCore and camera streams. 

In old versions of Filament (pre-1.22), the Stream.builder() would directly take a generated GL textureID which we would also pass to ARCore's setTextureName(), allowing ARCore to render frames to the Texture via Stream:
```
        Stream stream =
                new Stream.Builder()
                        .stream(textureId)
                        .width(width)
                        .height(height)
                        .build(engine);

        filamentTexture =
                new com.google.android.filament.Texture.Builder()
                        .sampler(Texture.Sampler.SAMPLER_EXTERNAL)
                        .format(Texture.InternalFormat.RGB8)
                        .build(engine);

        filamentTexture.setExternalStream(
                engine,
                stream);
```

In current versions of Filament, I'm not seeing a way to use Stream anymore with ARCore's camera as Stream now requires a SurfaceTexture. 

So is the use of Stream completely deprecated for this scenario? Should I instead use multiple textures and pass them to ARCore's setTextureNames()? Is this the recommended approach?

**Comment by poweifeng** [comment]
> Hi, I'm facing the same problem using ARCore and camera streams.
> 
> In old versions of Filament (pre-1.22), the Stream.builder() would directly take a generated GL textureID which we would also pass to ARCore's setTextureName(), allowing ARCore to render frames to the Texture via Stream:
> 
> ```
>         Stream stream =
>                 new Stream.Builder()
>                         .stream(textureId)
>                         .width(width)
>                         .height(height)
>                         .build(engine);
> 
>         filamentTexture =
>                 new com.google.android.filament.Texture.Builder()
>                         .sampler(Texture.Sampler.SAMPLER_EXTERNAL)
>                         .format(Texture.InternalFormat.RGB8)
>                         .build(engine);
> 
>         filamentTexture.setExternalStream(
>                 engine,
>                 stream);
> ```
> 
> In current versions of Filament, I'm not seeing a way to use Stream anymore with ARCore's camera as Stream now requires a SurfaceTexture.
> 
> So is the use of Stream completely deprecated for this scenario? Should I instead use multiple textures and pass them to ARCore's setTextureNames()? Is this the recommended approach?

this [example](https://github.com/google/filament/blob/main/android/samples/sample-hello-camera/src/main/java/com/google/android/filament/hellocam/CameraHelper.kt#L106) shows you how to use stream with ImageReader, maybe that can work for you?

**Comment by ss9326** [comment]
@poweifeng From my understanding, the example uses an acquired stream (hardware buffer), while I'm looking to build a native stream using a pre-generated GL Texture ID which is shared with ARCore. 

The hardware buffers approach is a no-go for me due to ARCore not supporting it for older Android APIs (below 27)
https://developers.google.com/ar/reference/java/com/google/ar/core/Config.TextureUpdateMode#EXPOSE_HARDWARE_BUFFER

So here's what I'm trying to do- 

1. Use glGenTextures to create a GL Texture and return a textureId

```
  public static int createCameraTexture() {
    int[] textures = new int[1];
    GLES30.glGenTextures(1, textures, 0);
    int result = textures[0];

    final int textureTarget = GLES11Ext.GL_TEXTURE_EXTERNAL_OES;
    GLES30.glBindTexture(textureTarget, result);
    GLES30.glTexParameteri(textureTarget, GLES30.GL_TEXTURE_WRAP_S, GLES30.GL_CLAMP_TO_EDGE);
    GLES30.glTexParameteri(textureTarget, GLES30.GL_TEXTURE_WRAP_T, GLES30.GL_CLAMP_TO_EDGE);
    GLES30.glTexParameteri(textureTarget, GLES30.GL_TEXTURE_MIN_FILTER, GLES30.GL_NEAREST);
    GLES30.glTexParameteri(textureTarget, GLES30.GL_TEXTURE_MAG_FILTER, GLES30.GL_NEAREST);
    return result;
  }
```

2. Pass the textureId to ARCore via setCameraTextureName so it can render its camera frames to the texture
```
  if (!hasSetTextureNames) {
      session.setCameraTextureName(cameraTextureId);
      hasSetTextureNames = true;
  }
```

3. Now, with pre-1.22 Filament, we could create a Stream using the same textureId-

```
  Stream stream =
          new Stream.Builder()
                  .stream(textureId)
                  .width(width)
                  .height(height)
                  .build(engine);

  filamentTexture =
          new com.google.android.filament.Texture.Builder()
                  .sampler(Texture.Sampler.SAMPLER_EXTERNAL)
                  .format(Texture.InternalFormat.RGB8)
                  .build(engine);

  filamentTexture.setExternalStream(
          engine,
          stream);
```

But the post-1.22 Stream API requires a SurfaceTexture in stream(), not textureId. I've also tried creating a SurfaceTexture with the textureId (or just 0) and passing that to the stream, but that results in a blank camera render. 

```
        SurfaceTexture surfaceTexture = new SurfaceTexture(textureId);

        filamentStream = new Stream.Builder()
                .stream(surfaceTexture)
                .width(width)
                .height(height)
                .build(EngineInstance.getEngine().getFilamentEngine());

        filamentTexture =
                new com.google.android.filament.Texture.Builder()
                        .width(width)
                        .height(height)
                        .sampler(Texture.Sampler.SAMPLER_EXTERNAL)
                        .format(Texture.InternalFormat.RGB8)
                        .build(engine);

        filamentTexture.setExternalStream(
                engine,
                filamentStream);
```

If I skip building the Stream entirely, and simply build a texture with importTexture(textureId), that works, but then I get the problem mentioned in the OP.

So I'm unsure of what the alternative is to make the Stream work properly with my pre-generated textureId.

The OP (authors of SceneView) solved this by NOT using a Stream, and instead using multiple (5+) textures, presumably so there's less chance of "texture that is being sampled from is also one that is being written to by camera" (pointed out above by @pixelflinger ).
https://github.com/SceneView/sceneview-android/blob/main/arsceneview/src/main/java/io/github/sceneview/ar/camera/ARCameraStream.kt

My question is - Is the solution adopted by the OP the correct way to fix this? Or is there a way to make the Stream work in this scenario (without using hardware buffers)?

Would really appreciate some insight on this, thanks.


EDIT: Seems the wave effect on my side was caused by Renderer's DynamicResolutionOptions being enabled on newer Filament versions. importTexture() works fine without Stream. I'd still be interested to know if Stream is useful/meant to be used in this case though.
